{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b64370d0",
   "metadata": {},
   "source": [
    "### Q1. Explain why we have to use the Exception class while creating a Custom Exception.   (Note: Here Exception class refers to the base class for all the exceptions.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9de61830",
   "metadata": {},
   "source": [
    "Ans)In Python, the built-in Exception class is the base class for all exceptions. When you create a custom exception, you inherit from Exception to ensure that your custom error behaves like other built-in exceptions and integrates properly with Python’s exception handling system.\n",
    "\n",
    "By subclassing Exception, your custom exception becomes part of the exception hierarchy, making it compatible with try-except blocks and allowing it to be caught specifically using except clauses. This enhances code readability and precision in error handling, as developers can handle particular exceptions without capturing unrelated ones.Therefore, using the Exception class as a base ensures that your custom exceptions are both functional and consistent with Python’s overall design."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ea0f5b",
   "metadata": {},
   "source": [
    "### Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad465d01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python Exception Hierarchy:\n",
      "BaseException\n",
      "    Exception\n",
      "        TypeError\n",
      "            MultipartConversionError\n",
      "            FloatOperation\n",
      "        StopAsyncIteration\n",
      "        StopIteration\n",
      "        ImportError\n",
      "            ModuleNotFoundError\n",
      "            ZipImportError\n",
      "        OSError\n",
      "            ConnectionError\n",
      "                BrokenPipeError\n",
      "                ConnectionAbortedError\n",
      "                ConnectionRefusedError\n",
      "                ConnectionResetError\n",
      "                    RemoteDisconnected\n",
      "            BlockingIOError\n",
      "            ChildProcessError\n",
      "            FileExistsError\n",
      "            FileNotFoundError\n",
      "            IsADirectoryError\n",
      "            NotADirectoryError\n",
      "            InterruptedError\n",
      "                InterruptedSystemCall\n",
      "            PermissionError\n",
      "            ProcessLookupError\n",
      "            TimeoutError\n",
      "            UnsupportedOperation\n",
      "            herror\n",
      "            gaierror\n",
      "            timeout\n",
      "            Error\n",
      "                SameFileError\n",
      "            SpecialFileError\n",
      "            ExecError\n",
      "            ReadError\n",
      "            SSLError\n",
      "                SSLCertVerificationError\n",
      "                SSLZeroReturnError\n",
      "                SSLWantReadError\n",
      "                SSLWantWriteError\n",
      "                SSLSyscallError\n",
      "                SSLEOFError\n",
      "            URLError\n",
      "                HTTPError\n",
      "                ContentTooShortError\n",
      "            BadGzipFile\n",
      "        EOFError\n",
      "            IncompleteReadError\n",
      "        RuntimeError\n",
      "            RecursionError\n",
      "            NotImplementedError\n",
      "                ZMQVersionError\n",
      "                StdinNotImplementedError\n",
      "            _DeadlockError\n",
      "            BrokenBarrierError\n",
      "            BrokenExecutor\n",
      "            SendfileNotAvailableError\n",
      "            ExtractionError\n",
      "            VariableError\n",
      "        NameError\n",
      "            UnboundLocalError\n",
      "        AttributeError\n",
      "            FrozenInstanceError\n",
      "        SyntaxError\n",
      "            IndentationError\n",
      "                TabError\n",
      "        LookupError\n",
      "            IndexError\n",
      "            KeyError\n",
      "                NoSuchKernel\n",
      "                UnknownBackend\n",
      "            CodecRegistryError\n",
      "        ValueError\n",
      "            UnicodeError\n",
      "                UnicodeEncodeError\n",
      "                UnicodeDecodeError\n",
      "                UnicodeTranslateError\n",
      "            UnsupportedOperation\n",
      "            JSONDecodeError\n",
      "            SSLCertVerificationError\n",
      "            Error\n",
      "            ClassNotFound\n",
      "            ClipboardEmpty\n",
      "            MessageDefect\n",
      "                NoBoundaryInMultipartDefect\n",
      "                StartBoundaryNotFoundDefect\n",
      "                CloseBoundaryNotFoundDefect\n",
      "                FirstHeaderLineIsContinuationDefect\n",
      "                MisplacedEnvelopeHeaderDefect\n",
      "                MissingHeaderBodySeparatorDefect\n",
      "                MultipartInvariantViolationDefect\n",
      "                InvalidMultipartContentTransferEncodingDefect\n",
      "                UndecodableBytesDefect\n",
      "                InvalidBase64PaddingDefect\n",
      "                InvalidBase64CharactersDefect\n",
      "                InvalidBase64LengthDefect\n",
      "                HeaderDefect\n",
      "                    InvalidHeaderDefect\n",
      "                    HeaderMissingRequiredValue\n",
      "                    NonPrintableDefect\n",
      "                    ObsoleteHeaderDefect\n",
      "                    NonASCIILocalPartDefect\n",
      "            IllegalMonthError\n",
      "            IllegalWeekdayError\n",
      "            MacroToEdit\n",
      "            ParserError\n",
      "            InvalidFileException\n",
      "            InvalidVersion\n",
      "            InvalidSpecifier\n",
      "            InvalidMarker\n",
      "            UndefinedComparison\n",
      "            UndefinedEnvironmentName\n",
      "            InvalidRequirement\n",
      "                RequirementParseError\n",
      "        AssertionError\n",
      "        ArithmeticError\n",
      "            FloatingPointError\n",
      "            OverflowError\n",
      "            ZeroDivisionError\n",
      "                DivisionByZero\n",
      "                DivisionUndefined\n",
      "            DecimalException\n",
      "                Clamped\n",
      "                Rounded\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Inexact\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Subnormal\n",
      "                    Underflow\n",
      "                DivisionByZero\n",
      "                FloatOperation\n",
      "                InvalidOperation\n",
      "                    ConversionSyntax\n",
      "                    DivisionImpossible\n",
      "                    DivisionUndefined\n",
      "                    InvalidContext\n",
      "        SystemError\n",
      "            CodecRegistryError\n",
      "        ReferenceError\n",
      "        MemoryError\n",
      "        BufferError\n",
      "        Warning\n",
      "            UserWarning\n",
      "                GetPassWarning\n",
      "                FormatterWarning\n",
      "            DeprecationWarning\n",
      "                ProvisionalWarning\n",
      "            PendingDeprecationWarning\n",
      "            SyntaxWarning\n",
      "            RuntimeWarning\n",
      "                ProactorSelectorThreadWarning\n",
      "                UnknownTimezoneWarning\n",
      "                PEP440Warning\n",
      "            FutureWarning\n",
      "                ProvisionalCompleterWarning\n",
      "            ImportWarning\n",
      "            UnicodeWarning\n",
      "            BytesWarning\n",
      "            ResourceWarning\n",
      "            DeprecatedTzFormatWarning\n",
      "            PkgResourcesDeprecationWarning\n",
      "        Error\n",
      "        _OptionError\n",
      "        error\n",
      "        Verbose\n",
      "        _Error\n",
      "        SubprocessError\n",
      "            CalledProcessError\n",
      "            TimeoutExpired\n",
      "        _GiveupOnSendfile\n",
      "        error\n",
      "        LZMAError\n",
      "        RegistryError\n",
      "        _GiveupOnFastCopy\n",
      "        ZMQBaseError\n",
      "            ZMQError\n",
      "                ContextTerminated\n",
      "                Again\n",
      "                InterruptedSystemCall\n",
      "            ZMQBindError\n",
      "            NotDone\n",
      "        Error\n",
      "        error\n",
      "        PickleError\n",
      "            PicklingError\n",
      "            UnpicklingError\n",
      "        _Stop\n",
      "        TokenError\n",
      "        StopTokenizing\n",
      "        ClassFoundException\n",
      "        EndOfBlock\n",
      "        TraitError\n",
      "        ArgumentError\n",
      "        ArgumentTypeError\n",
      "        ConfigError\n",
      "            ConfigLoaderError\n",
      "                ArgumentError\n",
      "            ConfigFileNotFound\n",
      "        ConfigurableError\n",
      "            MultipleInstanceError\n",
      "        ApplicationError\n",
      "        NoIPAddresses\n",
      "        Error\n",
      "            CancelledError\n",
      "            TimeoutError\n",
      "            InvalidStateError\n",
      "        Incomplete\n",
      "        TimeoutError\n",
      "        InvalidStateError\n",
      "        LimitOverrunError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        InvalidPortNumber\n",
      "        Empty\n",
      "        Full\n",
      "        DuplicateKernelError\n",
      "        TimeoutError\n",
      "        COMError\n",
      "        ArgumentError\n",
      "        ReturnValueIgnoredError\n",
      "        KeyReuseError\n",
      "        UnknownKeyError\n",
      "        LeakedCallbackError\n",
      "        BadYieldError\n",
      "        ReturnValueIgnoredError\n",
      "        Return\n",
      "        ErrorDuringImport\n",
      "        BdbQuit\n",
      "        Restart\n",
      "        FindCmdError\n",
      "        HomeDirError\n",
      "        ProfileDirError\n",
      "        IPythonCoreError\n",
      "            TryNext\n",
      "            UsageError\n",
      "            StdinNotImplementedError\n",
      "        InputRejected\n",
      "        GetoptError\n",
      "        OptionError\n",
      "        ErrorToken\n",
      "        PrefilterError\n",
      "        AliasError\n",
      "            InvalidAliasError\n",
      "        Error\n",
      "            InterfaceError\n",
      "            DatabaseError\n",
      "                InternalError\n",
      "                OperationalError\n",
      "                ProgrammingError\n",
      "                IntegrityError\n",
      "                DataError\n",
      "                NotSupportedError\n",
      "        Warning\n",
      "        SpaceInInput\n",
      "        DOMException\n",
      "            IndexSizeErr\n",
      "            DomstringSizeErr\n",
      "            HierarchyRequestErr\n",
      "            WrongDocumentErr\n",
      "            InvalidCharacterErr\n",
      "            NoDataAllowedErr\n",
      "            NoModificationAllowedErr\n",
      "            NotFoundErr\n",
      "            NotSupportedErr\n",
      "            InuseAttributeErr\n",
      "            InvalidStateErr\n",
      "            SyntaxErr\n",
      "            InvalidModificationErr\n",
      "            NamespaceErr\n",
      "            InvalidAccessErr\n",
      "            ValidationErr\n",
      "        ValidationError\n",
      "        EditReadOnlyBuffer\n",
      "        _Retry\n",
      "        InvalidLayoutError\n",
      "        HeightIsUnknownError\n",
      "        ParserSyntaxError\n",
      "        InternalParseError\n",
      "        _PositionUpdatingFinished\n",
      "        SimpleGetItemNotFound\n",
      "        UncaughtAttributeError\n",
      "        ParamIssue\n",
      "        _JediError\n",
      "            InternalError\n",
      "            WrongVersion\n",
      "            RefactoringError\n",
      "        OnErrorLeaf\n",
      "        InvalidPythonEnvironment\n",
      "        MessageError\n",
      "            MessageParseError\n",
      "                HeaderParseError\n",
      "                BoundaryError\n",
      "            MultipartConversionError\n",
      "            CharsetError\n",
      "        Error\n",
      "        HTTPException\n",
      "            NotConnected\n",
      "            InvalidURL\n",
      "            UnknownProtocol\n",
      "            UnknownTransferEncoding\n",
      "            UnimplementedFileMode\n",
      "            IncompleteRead\n",
      "            ImproperConnectionState\n",
      "                CannotSendRequest\n",
      "                CannotSendHeader\n",
      "                ResponseNotReady\n",
      "            BadStatusLine\n",
      "                RemoteDisconnected\n",
      "            LineTooLong\n",
      "        InteractivelyDefined\n",
      "        KillEmbedded\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        ExpatError\n",
      "        Error\n",
      "            ProtocolError\n",
      "            ResponseError\n",
      "            Fault\n",
      "        BadZipFile\n",
      "        LargeZipFile\n",
      "        error\n",
      "        com_error\n",
      "        internal_error\n",
      "        ParseBaseException\n",
      "            ParseException\n",
      "            ParseFatalException\n",
      "                ParseSyntaxException\n",
      "        RecursiveGrammarException\n",
      "        ResolutionError\n",
      "            VersionConflict\n",
      "                ContextualVersionConflict\n",
      "            DistributionNotFound\n",
      "            UnknownExtra\n",
      "        UnableToResolveVariableException\n",
      "        InvalidTypeInArgsException\n",
      "    GeneratorExit\n",
      "    SystemExit\n",
      "    KeyboardInterrupt\n",
      "    CancelledError\n",
      "    DebuggerInitializationError\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(cls, indent=0):\n",
    "    print(' ' * indent + cls.__name__)\n",
    "    for subclass in cls.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "\n",
    "try:\n",
    "    print(\"Python Exception Hierarchy:\")\n",
    "    print_exception_hierarchy(BaseException)\n",
    "except Exception:\n",
    "    print(\"An error occurred while printing the exception hierarchy.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92d0bf19",
   "metadata": {},
   "source": [
    "### Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a270691d",
   "metadata": {},
   "source": [
    "In Python, the ArithmeticError class is a built-in base class for all errors that occur in numeric calculations. It is a subclass of the Exception class and serves as a parent for more specific arithmetic-related error types.\n",
    "\n",
    "The main errors defined under the ArithmeticError class are:\n",
    "\n",
    "ZeroDivisionError\n",
    "\n",
    "OverflowError\n",
    "\n",
    "FloatingPointError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d3b63261",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ZeroDivisionError occurred: division by zero\n"
     ]
    }
   ],
   "source": [
    "#1.ZeroDivisionError\n",
    "#This error is raised when a number is divided by zero, which is mathematically undefined.\n",
    "try:\n",
    "    result = 10 / 0\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"ZeroDivisionError occurred:\", e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a9826c79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OverflowError occurred: math range error\n"
     ]
    }
   ],
   "source": [
    "#2.OverflowError\n",
    "#This error occurs when the result of an arithmetic operation is too large to be represented within the available numeric type.\n",
    "import math\n",
    "\n",
    "try:\n",
    "    result = math.exp(1000)  # Exponential of a large number\n",
    "except OverflowError as e:\n",
    "    print(\"OverflowError occurred:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877af855",
   "metadata": {},
   "source": [
    "### Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3554f908",
   "metadata": {},
   "source": [
    "**The LookupError** class in Python is a base class for exceptions that occur when trying to access an element in a sequence or mapping (like lists, tuples, dictionaries) using an invalid key or index. It doesn't occur by itself but is inherited by more specific exceptions such as IndexError and KeyError. \n",
    "\n",
    "Using LookupError allows developers to catch a broad category of data-access errors in a generalized way, especially useful when the specific type of lookup failure (index vs. key) doesn’t matter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "310dca15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KeyError occurred: 'age'\n"
     ]
    }
   ],
   "source": [
    "#KeyError,This error occurs when trying to access a dictionary with a key that doesn’t exist.\n",
    "try:\n",
    "    my_dict = {'name': 'Alice'}\n",
    "    print(my_dict['age'])\n",
    "except KeyError as e:\n",
    "    print(\"KeyError occurred:\", e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b1cc86df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IndexError occurred: list index out of range\n"
     ]
    }
   ],
   "source": [
    "#IndexError,This error happens when trying to access a list element at an index that is out of range.\n",
    "try:\n",
    "    my_list = [1, 2, 3]\n",
    "    print(my_list[5])\n",
    "except IndexError as e:\n",
    "    print(\"IndexError occurred:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de39c976",
   "metadata": {},
   "source": [
    "### Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b759bf85",
   "metadata": {},
   "source": [
    "**ImportError** is a built-in Python exception that occurs when an import statement fails to import a module or an object from a module. This typically happens when:\n",
    "\n",
    "1.The module you're trying to import doesn't exist.\n",
    "\n",
    "2.There's a typo in the module name or object name.\n",
    "\n",
    "3.The module exists but has an internal error preventing it from loading.\n",
    "\n",
    "4.You're trying to import a class, function, or variable that doesn’t exist in the module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f913ba17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ImportError occurred: cannot import name 'square' from 'math' (unknown location)\n"
     ]
    }
   ],
   "source": [
    "#ImportError\n",
    "try:\n",
    "    from math import square  # 'square' doesn't exist in the math module\n",
    "except ImportError as e:\n",
    "    print(\"ImportError occurred:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fed87575",
   "metadata": {},
   "source": [
    "**ModuleNotFoundError** is a subclass of ImportError, introduced in Python 3.6, which specifically occurs when the module itself cannot be found. This makes it more specific than ImportError and helps in better error handling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "53deb0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ModuleNotFoundError occurred: No module named 'nonexistentmodule'\n"
     ]
    }
   ],
   "source": [
    "#ModuleError\n",
    "try:\n",
    "    import nonexistentmodule\n",
    "except ModuleNotFoundError as e:\n",
    "    print(\"ModuleNotFoundError occurred:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b65937f",
   "metadata": {},
   "source": [
    "### Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2007e59e",
   "metadata": {},
   "source": [
    "#### Here are some of the best practices for exception handling in Python:\n",
    "\n",
    "**1. Use Specific Exceptions:**\n",
    "Always catch specific exceptions rather than using a generic except: block. This ensures you don’t unintentionally catch unrelated errors.\n",
    "\n",
    "**2. Avoid Using Bare except:**\n",
    "Catching all exceptions without specifying the type can hide bugs and make debugging harder.\n",
    "\n",
    "**3. Use finally to Clean Up Resources:**\n",
    "Use finally to ensure certain cleanup actions (like closing files or releasing resources) are always executed.\n",
    "\n",
    "**4. Use else for Code That Runs Only If No Exception Occurs:**\n",
    "The else block can be used to separate normal code from the exception-handling logic.\n",
    "\n",
    "**5. Don’t Suppress Exceptions Silently:**\n",
    "Avoid empty except blocks that hide errors. Always handle or log the exception properly.\n",
    "\n",
    "**6. Raise Exceptions with Meaningful Messages:**\n",
    "When raising custom exceptions, include clear messages to help with debugging.\n",
    "\n",
    "**7. Create Custom Exceptions When Appropriate:**\n",
    "Use custom exceptions to represent application-specific errors.\n",
    "\n",
    "**8. Log Exceptions in Real Applications:**\n",
    "In larger applications, log exceptions using the logging module instead of printing them to the console.\n",
    "\n",
    "#### Using these best practices makes your code safer, easier to debug, and more maintainable in the long run.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
